app/api/controllers/validators/resolvers/Resolvers.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package api.controllers.validators.resolvers
+
+import api.models.errors.MtdError
+import cats.data.Validated
+import cats.data.Validated.Valid
+import cats.implicits._
+
+import scala.math.Ordered.orderingToOrdered
+
+trait Resolvers {
+  type SimpleResolver[In, Out] = In => Validated[Seq[MtdError], Out]
+  type Validator[A]            = A => Option[Seq[MtdError]]
+
+  implicit class SimpleResolverOps[In, Out](resolver: In => Validated[Seq[MtdError], Out]) {
+    def map[Out2](f: Out => Out2): SimpleResolver[In, Out2] = i => resolver(i).map(f)
+
+    def thenValidate(validator: Validator[Out]): SimpleResolver[In, Out] = i => resolver(i).andThen(o => validator(o).toInvalid(o))
+
+    def resolveOptionally: SimpleResolver[Option[In], Option[Out]] = _.map(in => resolver(in).map(Some(_))).getOrElse(Valid(None))
+
+    def resolveOptionallyWithDefault(default: => Out): SimpleResolver[Option[In], Out] = _.map(in => resolver(in)).getOrElse(Valid(default))
+  }
+
+  implicit class ValidatorOps[A](validator: A => Option[Seq[MtdError]]) {
+    def thenValidate(other: Validator[A]): Validator[A] = a => validator(a).orElse(other(a))
+  }
+
+  def satisfies[A](error: => MtdError)(predicate: A => Boolean): Validator[A] =
+    a => Option.when(!predicate(a))(List(error))
+
+  def inRange[A: Ordering](minAllowed: A, maxAllowed: A, error: => MtdError): Validator[A] =
+    satisfiesMin[A](minAllowed, error) thenValidate satisfiesMax[A](maxAllowed, error)
+
+  def satisfiesMin[A: Ordering](minAllowed: A, error: => MtdError): Validator[A] = satisfies(error)(minAllowed <= _)
+  def satisfiesMax[A: Ordering](maxAllowed: A, error: => MtdError): Validator[A] = satisfies(error)(_ <= maxAllowed)
+
+  def combinedValidator[A](first: Validator[A], others: Validator[A]*): Validator[A] = { value: A =>
+    val validators = first +: others
+
+    val validations = validators.map(validator => validator(value))
+
+    validations.reduce(_ combine _)
+  }
+
+}
+
+object Resolvers extends Resolvers
```

Comment:
Can be deleted?

Reply:
Just there to permit an import based alternative that can be used more locally if required.